### PR TITLE
executable crashing fix in progress, not yet working

### LINKED ISSUE #213: executable instantly crashes
The executable created by the 'create-executable' workflow instantly crashes because it does not have access to certain files and images throughout the project, mostly located in the '/shared' directory.

Either 'main.py' and/or 'create-executable.yml' need to be changed so that the resulting executable is functional on other machines.

What is required for this issue to be done:
- [ ] Github workflow correctly builds a windows executable that open to the Home page of Mouser

361553909-dea874fc-367a-4df4-b0e1-dbb7cae83614

### PR DESCRIPTION
**Pull request by ikardasevic**
Fixes #213

**What was changed?**

Changes have been made as comments as the fix is not yet functioning. But changes include adding import sys and import os to support dynamic path resolution. The new helper function get_resource_path() was added to dynamically resolve resource paths. The path to the image "shared/images/MouseLogo.png" was changed to use the get_resource_path() function instead of a static path. 

**Why was it changed?**

Adding import sys and import os to support dynamic path resolution for files, both in development and when the program is packaged as an executable. The new helper function was added to resolve resource paths, allowing the program to correctly locate assets whether its running as an executable or from source. The path was updated to use the new helper function instead of a static relative path, ensuring the image can be found regardless of how the program is run. 

**How was it changed?**

The file that was modified was main.py. Two lines were added at the beginning, a new function after the global variables, and a path change towards the end of the file.

### CODE REVIEW

**Comment by jackbelyeu** [comment]
I think some files got into your commits that are unnecessary. We shouldn't be changing the pull request template file in the repo, that would change all future pull requests. Do we need .DS_Store and the added Mouser submodule?

**Comment by jackbelyeu** [comment]
I like your comments and ideas for main.py. Try them and see if they work. You may also want to look at ```/.github/workflows/create-executable.yaml```, which defines the workflow you're working on. It may be possible to edit which files the executable has access to from here.

**Comment by ikardasevic** [comment]
Ok, I will look into the workflow you suggested. As for the submodule, it got added when I cloned. I thought I removed it, but it will be fixed by Monday. I will let you know if I run into any troubles. Thanks!

**Comment by ikardasevic** [comment]
Fixes https://github.com/oss-slu/Mouser/issues/213

What was changed?

I added os.path.abspath(filepath) at the beginning of each function to make sure any file paths passed as arguments are fully resolved to absolute paths. This should make sure paths are always fully resolved to absolute paths, we reduce the chances of encountering file-not-found errors due to working directory issues. The path to serial_port_preference.csv is now constructed dynamically using os.path.join(). I added a try-except block to catch FileNotFoundError if the file cannot be found

Why was it changed?

Dynamic construction of file paths dynamically to ensure consistent path handling across the program, and any other files that reference other external resources.


How was it changed?

The files that were modified was file_utils.py and serial_port_controller.py. I added os.path.abspath(filepath) at the beginning of each function to make sure any file paths passed as arguments are fully resolved to absolute paths in file_utils.py. Changes to serial_port_controller.py were merged and altered.

File: shared/serial_port_controller.py [hunk 1]
```diff
@@ -10,47 +10,29 @@
...
-                    self.retrieve_setting([line[0], line[3], line[1], line[4], line[2]])
-                    # Whenever automation is implemented, add the serial port here
-                    # self.set_reader_port(line[6])
-
-            # settings = [baud rate, data bits/byte size, parity, stop bits, flow control option]
-            # file format: baud rate, parity, flow control option, data bits, stop bits, input method (not needed), port
-
-
-
+            # Dynamically construct the file path
```
Comment (moshernoah) [inline shared/serial_port_controller.py:29]: This is a much more elegant solution to finding the absolute paths for the local files. Looks good!

**Comment by jackbelyeu** [comment]
🤯 BOOM! The .exe created by this launched and functioned! I'm going to push a commit that removes the MacOS build before merging for the following reasons:
- The client specifically uses Windows machines, so the Mac executable is unnecessary
- The MacOS artifact requires more development to function anyway
- The MacOS GitHub workflow sometimes fails resulting in a red x when the commit is good

I'm just going to edit the workflow to remove macOS, your code and functionality look great, thank you!

**Review by jackbelyeu — APPROVED** [review]

**Pull request merged**